yjit/src/codegen.rs

```diff
@@ -1389,24 +1389,25 @@ mod tests {
     }
 }
 
-/*
 // new array initialized from top N values
 fn gen_newarray(jit: &mut JITState, ctx: &mut Context, cb: &mut CodeBlock, ocb: &mut OutlinedCb) -> CodegenStatus
 {
-    rb_num_t n = (rb_num_t)jit_get_arg(jit, 0);
+    let n = jit_get_arg(jit, 0).as_u32();
 
     // Save the PC and SP because we are allocating
-    jit_prepare_routine_call(jit, ctx, REG0);
+    jit_prepare_routine_call(jit, ctx, cb, REG0);
 
-    let values_ptr = ctx.sp_opnd(-(SIZEOF_VALUE * (uint32_t)n));
+    let offset_magnitude = SIZEOF_VALUE as u32 * n;
+    let values_ptr = ctx.sp_opnd(-(offset_magnitude as isize));
 
     // call rb_ec_ary_new_from_values(struct rb_execution_context_struct *ec, long n, const VALUE *elts);
     mov(cb, C_ARG_REGS[0], REG_EC);
-    mov(cb, C_ARG_REGS[1], imm_opnd(n));
+    mov(cb, C_ARG_REGS[1], imm_opnd(n.into()));
     lea(cb, C_ARG_REGS[2], values_ptr);
-    call_ptr(cb, REG0, (void *)rb_ec_ary_new_from_values);
+    let ary_new = CodePtr::from(rb_ec_ary_new_from_values as *mut u8);
+    call_ptr(cb, REG0, ary_new);
```

Comment:
I think we might want to change `call_ptr` to take a function pointer type directly (or a generic function pointer type if there's such a thing) because we keep having to wrap the function pointers in a CodePtr. We can save that for another PR, but just noting, if we keep doing things that are unergonomic, we might want to change the API.

Reply:
That's fair. Maybe a call_ptr_raw that takes a * mut u8?

Reply:
I think we maybe should just make call_ptr take a `*const u8`. I don't think we ever really need to call it with a `CodePtr`.

Reply:
Or even make it take a `usize`. call_ptr should never need to read from the pointer.